13167 docs for connector builder new UI schemaform rollout (#63329)
airbytehq/airbyte: docs/platform/connector-development/connector-builder-ui/error-handling.md

> ### Wait time from header

Reviewer: ```suggestion
### Wait Time Extracted from Response Header
```

> The "Wait time from header" backoff strategy allows the connector to wait before retrying a request based on the value specified in the API response header.

Reviewer: ```suggestion
The "Wait Time Extracted from Response Header" backoff strategy allows the connector to wait before retrying a request based on the value specified in the API response header.
```

> ### Wait until time from header

Reviewer: ```suggestion
### Wait Until Time Defined in Response Header
```

> The "Wait until time from header" backoff strategy allows the connector to wait until a specific time before retrying a request according to the API response header. This strategy extracts a timestamp from the response header and waits until that time before retrying.

Reviewer: ```suggestion
The "Wait Until Time Defined in Response Header" backoff strategy allows the connector to wait until a specific time before retrying a request according to the API response header. This strategy extracts a timestamp from the response header and waits until that time before retrying.
```

## Response Filters

A response filter should be used when a connector needs to interpret an API response to decide how the sync should proceed. Response filters allow you to define specific conditions that determine whether a response should be retried, ignored, treated as successful, cause the sync to fail, or be treated as rate-limited.

Multiple response filters can be configured for a single error handler. When using multiple filters, they are applied sequentially and the response will be processed according to the first filter that matches.

<iframe width="640" height="716" src="https://www.loom.com/embed/dc86147384204156a2b79442a00c0dd3" frameborder="0" webkitallowfullscreen mozallowfullscreen allowfullscreen></iframe>

### Response Conditions

The following conditions can be specified on response filters to determine if attributes of the response match the filter. When more than one condition is specified, the filter will take action if the response satisfies any of the conditions:

> - [Error message contains](#error-message-contains)

Reviewer: ```suggestion
- [Error Message Substring](#error-message-substring)
```

> - [Predicate evaluation](#predicate-evaluation)

Reviewer: ```suggestion
- [Predicate](#predicate)
```

> - [HTTP status codes](#http-status-codes)

Reviewer: ```suggestion
- [HTTP Codes](#http-codes)
```

> #### Error message contains

Reviewer: ```suggestion
#### Error Message Substring
```

> For a response filter that defines the "Error message contains" field, the connector will check if the provided text exists within the text body of the API response. If the text is present, the response filter will carry out the specified action.

Reviewer: ```suggestion
For a response filter that defines the "Error Message Substring" field, the connector will check if the provided text exists within the text body of the API response. If the text is present, the response filter will carry out the specified action.
```

**Configuration:**
- **Error message substring**: The text to search for in the response body

**Example:**
- Value: `"This API operation is not enabled for this site"`

##### Example

For the Chargebee API, some endpoints are only available for a specific API version and if an endpoint is unavailable, the response text will contain `"This API operation is not enabled for this site"`. The Airbyte Chargebee integration allows customers to configure which API version to use when retrieving data for a stream. When the connector makes requests to Chargebee using an unsupported version, the response filter will match according to the response text and proceeds based on the configured action.

> #### Predicate evaluation

Reviewer: ```suggestion
#### Predicate
```

This field allows for more granular control over how the response filter matches against attributes of an API response. The predicate is an interpolation expression that is evaluated against the API response's text body or headers.

**Configuration:**
- **Predicate**: An interpolation expression that evaluates to true or false based on the response content

**Examples:**
- `"{{ 'Too much requests' in response }}"` (check if text exists in response)
- `"{{ response.code == 300 }}"` (check specific field value)
- `"{{ 'error_code' in response and response['error_code'] == 'ComplexityException' }}"` (complex condition)

##### Example

For the Zoom API, the response text body can include a special non-error status codes under the `code` field. An example response text body would look like `{"code": 300}`. The "Error message contains" condition is too broad because there could be record data containing the text "300". Instead, for a response filter defining the predicate as `{{ response.code == 300 }}`, during a sync, the predicate expression will be evaluated to true and the connector proceeds based on the configured action.

> #### HTTP status codes

Reviewer: ```suggestion
#### HTTP Codes
```

A response filter can specify a set of numeric HTTP status codes to match against. When receiving an API response, the connector will check if the status code of the response is in the provided set of HTTP status codes.

**Configuration:**
- **HTTP codes**: An array of HTTP status codes to match against

**Examples:**
- `[420, 429]` (match rate limiting codes)
- `[500]` (match server errors)
- `[403, 404]` (match specific client errors)

##### Example

The Pocket API emits API responses for rate limiting errors using a 403 error status code. The default error handler interprets 403 errors as non-retryable and will fail the sync when they are encountered. The connector can configure a response filter with HTTP status codes that contains 403. When a 403 error response from the API is encountered, the connector proceeds based on the configured action.

> ### Response Actions

Reviewer: ```suggestion
### Action
```

> ### Response Actions

Reviewer: We may also want to put the Action and Failure types sections above the other ones, to match the order they are shown in the UI

### Failure Types

When using the FAIL action, you can optionally specify a failure type to categorize the error:

- **system_error**: Indicates a system-level error that is not related to user configuration
- **config_error**: Indicates an error caused by incorrect user configuration
- **transient_error**: Indicates a temporary error that might resolve itself

### Custom Error Messages

The "Error message" field allows you to customize the message that is relayed back to users when the API response matches a response filter. This field supports interpolation, allowing you to include dynamic information from the response in the error message.

**Configuration:**
- **Error message**: Custom error message to display when the filter matches

**Interpolation context available:**
- `config`: Access to connector configuration values
- `response`: Access to the API response content
- `headers`: Access to response headers

**Examples:**
- `"API rate limit exceeded. Please try again later."`
- `"Invalid API key: {{ response.error_message }}"`
- `"Request failed with status {{ response.status_code }}: {{ response.message }}"`

## Advanced Error Handling Configuration

> ### Max Retries

Reviewer: ```suggestion
### Max Retry Count
```

> The "Max retries" field allows you to configure the maximum number of times a request will be retried before giving up. This applies to all retryable responses (those with RETRY or RATE_LIMITED actions).

Reviewer: ```suggestion
The "Max retry count" field allows you to configure the maximum number of times a request will be retried before giving up. This applies to all retryable responses (those with RETRY or RATE_LIMITED actions).
```

**Configuration:**
- **Max retry count**: Maximum number of retry attempts (default: 5)

**Examples:**
- `5` (default - retry up to 5 times)
- `0` (no retries)
- `10` (retry up to 10 times for persistent issues)

> ### Multiple Error Handlers
> 
> In the "Error handlers" section of a stream, one or more handlers can be defined. When multiple error handlers are specified, the response will be evaluated against each error handler in the order they are defined. The connector will take the action of the first error handler that matches the response and ignore subsequent handlers.
> 
> This is particularly useful when you need different handling logic for different types of errors:
> 
> **Example scenarios:**
> - Handle rate limiting (429) differently from server errors (5XX)
> - Apply different retry strategies for different error codes
> - Use custom error messages for specific API error responses
> 

Reviewer: ```suggestion
```
This is covered by the `Composite Error Handler Usage` section below

### Composite Error Handler Usage

When using a Composite Error Handler, you can chain multiple error handlers together. Each error handler in the list is evaluated in sequence until one matches the response. This allows for sophisticated error handling workflows:

1. **First handler**: Check for specific API error codes and handle them with custom logic
2. **Second handler**: Handle rate limiting with appropriate backoff
3. **Third handler**: Default handler for all other errors

### Custom Error Handler Implementation

For advanced use cases that cannot be handled through configuration, you can implement a Custom Error Handler. This requires:

1. **Class name**: Fully-qualified name of your custom error handler class
2. **Implementation**: The class must implement the required error handler interface
3. **Parameters**: Optional parameters to pass to your custom implementation

**Example:**
- Class name: `"source_myapi.components.MyCustomErrorHandler"`

## Common Error Handling Patterns

### Rate Limiting

Most APIs implement rate limiting. Here's a recommended approach:

1. **Use appropriate backoff strategy**: 

>    - "Wait time from header" if the API provides `Retry-After` header

Reviewer: ```suggestion
   - "Wait Until Time Defined in Response Header" if the API provides `Retry-After` header
```
